### PR TITLE
[FEATURE] UI for creating / editing a shipping

### PR DESCRIPTION
**Pull request by turbo-ele**
**Tasks:**
- [x] ui for creating a shipping
- [x] convert shipping status via toolbar
- [x] implemented bitmask and new status handling
- [x] gather feedback for my changes

**Informations:**

| Q | A |
| --- | --- |
| Tests pass? | aye |
| Fixed tickets | - |
| BC Breaks | - |
| Doc | - |

**Requires:**

### CODE REVIEW

File: Admin/OrderContentNavigationExtension.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+use Sulu\Bundle\AdminBundle\Navigation\ContentNavigationInterface;
+use Sulu\Bundle\AdminBundle\Navigation\ContentNavigationItem;
+
+class OrderContentNavigationExtension implements ContentNavigationInterface
+{
+    private $navigation = array();
+
+    public function __construct()
+    {
+        $permissions = new ContentNavigationItem('salesshipping.shippings.title');
```
Comment (michaelzangerle) [inline Admin/OrderContentNavigationExtension.php]: @turbo-ele the variable name should probably be shipping?

File: Admin/SuluSalesShippingContentNavigation.php [hunk 2]
```diff
@@ -24,7 +24,7 @@ public function __construct()
 
         /* CONTACTS */
         // details
-        $overview = new ContentNavigationItem('public.overview');
+        $overview = new ContentNavigationItem('public.details');
         $overview->setAction('overview');
```
Comment (michaelzangerle) [inline Admin/SuluSalesShippingContentNavigation.php:28]: @turbo-ele overview should probably also be changed to details?

File: Api/Shipping.php [hunk 3]
```diff
@@ -19,6 +20,10 @@
 class Shipping extends ApiWrapper
 {
     /**
+     * @Exclude
+     */
+    private $shippingItems;
+    /**
```
Comment (michaelzangerle) [inline Api/Shipping.php:28]: @turbo-ele add a blank line here

File: Api/Shipping.php [hunk 4]
```diff
@@ -10,6 +10,7 @@
 use JMS\Serializer\Annotation\SerializedName;
 use Sulu\Component\Security\UserInterface;
 use Sulu\Bundle\Sales\ShippingBundle\Entity\Shipping as ShippingEntity;
+use JMS\Serializer\Annotation\Exclude;
 
 /**
  * The Shipping class which will be 2exported to the API
```
Comment (michaelzangerle) [inline Api/Shipping.php]: @turbo-ele typo

File: Api/Shipping.php [hunk 5]
```diff
@@ -471,12 +476,14 @@ public function removeShippingItem(\Sulu\Bundle\Sales\ShippingBundle\Entity\Ship
      */
     public function getItems()
     {
-        $items = array();
-        foreach ($this->entity->getShippingItems() as $shippingItem) {
-            $items[] = new ShippingItem($shippingItem, $this->locale);
+        if (!$this->shippingItems) {
+            $this->shippingItems = array();
+            foreach ($this->entity->getShippingItems() as $shippingItem) {
+                $this->shippingItems[] = new ShippingItem($shippingItem, $this->locale);
```
Comment (michaelzangerle) [inline Api/Shipping.php:484]: @turbo-ele no import for ShippingItem?

File: Api/ShippingItem.php [hunk 6]
```diff
@@ -130,4 +136,24 @@ public function getItem()
...
+     */
+    public function setShippedItems($numShippedItems)
+    {
+        $this->shippedItems = $numShippedItems;
+
+        return $this;
+    }
+
+    /**
+     * @return shippedItems*
```
Comment (michaelzangerle) [inline Api/ShippingItem.php]: @turbo-ele typo here and could be extended to "number of shipped items" to clarify what will be returned

File: Controller/ShippingController.php [hunk 7]
```diff
@@ -140,6 +162,17 @@ function ($id) use ($locale) {
     }
 
     /**
+     * @param Request $request
```
Comment (michaelzangerle) [inline Controller/ShippingController.php:167]: @turbo-ele add a better description about what to expect as response and/or a better name for the function

File: Entity/ShippingRepository.php [hunk 8]
```diff
@@ -58,6 +59,11 @@ public function findByLocaleAndFilter($locale, array $filter)
                         $qb->andWhere('status.id = :' . $key);
                         $qb->setParameter($key, $value);
                         break;
+                    case 'orderId':
+                        $qb->leftJoin('shipping.order', 'o');
+                        $qb->andWhere('o.id = :' . $key);
+                        $qb->setParameter($key, $value);
+                        break;
                 }
```
Comment (michaelzangerle) [inline Entity/ShippingRepository.php:67]: @turbo-ele would it make sense to define a default case here?

Comment (turbo-ele) [inline Entity/ShippingRepository.php:67]: no because default would be nothing at all

File: Api/Shipping.php (hunk 5 above)
Comment (turbo-ele) [inline Api/Shipping.php:484]: done, even though it's not necessary, since they are in the same folder

**Comment by michaelzangerle** [comment]
@turbo-ele there seem to be a few translation keys missig like for example navigation.sales.shipping and  public.details

File: Resources/public/js/components/orders/components/shippings-list/main.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+*/
+
+define(function() {
+
+    'use strict';
+
+    var bindCustomEvents = function() {
+            // navigate to edit contact
+            this.sandbox.on('husky.datagrid.item.click', function(item) {
+//                this.sandbox.emit('sulu.sidebar.set-widget', '/admin/widget-groups/contact-info?contact=' + item);
```
Comment (michaelzangerle) [inline Resources/public/js/components/orders/components/shippings-list/main.js:17]: @turbo-ele is the listener needed at all? format code in this file!

File: Resources/public/js/components/shippings/components/form/main.js [hunk 10]
```diff
@@ -1,607 +1,431 @@
...
+                    },
+                    divider: {
+                        divider: true
+                    }
+                };
+
+            // if shipping is not new
+            if (this.options.data.id) {
+                // define workflow based on status
+                // TODO: get statuses from backend
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/components/form/main.js]: @turbo-ele todo?

File: Resources/public/js/components/orders/components/shippings-list/main.js (hunk 9 above)
Comment (turbo-ele) [inline Resources/public/js/components/orders/components/shippings-list/main.js:17]: no, this is just a preparation for the sidebar. currently it's not needed at all

File: Resources/public/js/components/shippings/components/form/main.js [hunk 11]
```diff
@@ -1,607 +1,431 @@
...
+            });
+        },
+
+        /**
+         * checks for unsaved data. if unsaved, a dialog is shown, else immediately proceed
+         * @param callback - called when no unsaved data, or warning was confirmed
+         */
+        checkForUnsavedData = function(callback) {
+            if (typeof callback !== 'function') {
+                return;
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/components/form/main.js]: @turbo-ele to throw an exception or log an error would probably be a good idea at this point

File: Resources/public/js/components/shippings/components/form/main.js [hunk 12]
```diff
@@ -1,607 +1,431 @@
...
+            this.sandbox.on('sulu.header.toolbar.save', function() {
+                this.submit();
+            }, this);
+
+            // back to list
+            this.sandbox.on('sulu.header.back', function() {
+                this.sandbox.emit('sulu.salesshipping.shippings.list');
+            }, this);
+
+            // TODO expected deliverydate
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/components/form/main.js]: @turbo-ele what do you mean by this? is it still open?

File: Resources/public/js/components/shippings/components/form/main.js [hunk 13]
```diff
@@ -1,607 +1,431 @@
...
+                        '.changeListener input, ' +
+                        '.changeListener textarea');
+
+                // change in item-table
+                this.sandbox.on('sulu.item-table.changed', changeHandler.bind(this));
+
 //                // listen on
-//                this.sandbox.on('husky.select.contact-select.selected.item', changeHandler.bind(this));
-//                this.sandbox.on('husky.select.responsible-contact-select.selected.item', changeHandler.bind(this));
 //                this.sandbox.on('husky.select.delivery-terms.selected.item', changeHandler.bind(this));
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/components/form/main.js]: @turbo-ele update of save button after change of terms works already - so this listener is not needed anymore is he?

File: Resources/public/js/components/shippings/components/list/main.js [hunk 14]
```diff
@@ -40,16 +40,11 @@ define(function() {
                 rightSpace: false
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/components/list/main.js:40]: @turbo-ele format code

File: Resources/public/js/components/shippings/main.js [hunk 15]
```diff
@@ -19,169 +21,186 @@ define([
             this.bindCustomEvents();
             this.bindSidebarEvents();
             this.shipping = null;
+            this.order = null;
 
             if (this.options.display === 'list') {
                 this.renderList();
             } else if (this.options.display === 'form') {
                 this.renderForm().then(function() {
+//                    {
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/main.js]: @turbo-ele what about this code?

File: Resources/public/js/components/shippings/main.js [hunk 16]
```diff
@@ -19,169 +21,186 @@ define([
...
+            this.sandbox.on('sulu.salesshipping.shippings.list', this.listAction.bind(this));
+        },
+
+        /**
+         * Binds general sidebar events
+         */
+        bindSidebarEvents: function() {
+            // TODO: uncomment after sidebar is implemented
+            // bind sidebar
+//            this.sandbox.dom.off('#sidebar');
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/main.js]: @turbo-ele same here

File: Resources/public/js/components/shippings/main.js [hunk 17]
```diff
@@ -193,38 +212,103 @@ define([
...
             ]);
+        },
+
+        renderOrderList: function() {
+            var dfd = this.sandbox.data.deferred();
+            if (!!this.options.id) {
+                this.order = Order.findOrCreate({id: this.options.id});
+
+                this.order.fetch({
+                    success: function(model) {
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/main.js]: @turbo-ele model is unused

File: Resources/public/js/components/shippings/main.js [hunk 18]
```diff
@@ -193,38 +212,103 @@ define([
...
+                    }.bind(this),
+                    error: this.errorCallback.bind(this, dfd)
+                });
+
+                // order id is set
+            } else if (!!this.options.orderId) {
+                this.order = Order.findOrCreate({id: this.options.orderId});
+                this.shipping = this.shipping.set({order: this.order});
+
+                var shippedOrderItems = null,
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/main.js]: @turbo-ele too many var statements

File: Resources/public/js/components/shippings/main.js [hunk 19]
```diff
@@ -193,38 +212,103 @@ define([
...
+
+                this.sandbox.data.when(dfdOrder, dfdShipped).then(function() {
+                    this.startForm(dfd, this.shipping, shippedOrderItems);
+                }.bind(this));
+            }
+            return dfd.promise();
+        },
+
+        startForm: function(dfd, shipping, shippedOrderItems) {
+            var $form = this.sandbox.dom.createElement('<div id="shipping-form-container"/>')
```
Comment (michaelzangerle) [inline Resources/public/js/components/shippings/main.js]: @turbo-ele missing semicolon

File: Resources/public/js/main.js [hunk 20]
```diff
@@ -17,6 +18,10 @@ define({
 
     name: "SuluSalesShippingBundle",
 
+    STATUS_CREATED : 1,
```
Comment (michaelzangerle) [inline Resources/public/js/main.js]: @turbo-ele states already defined in util/shippingStatus.js? Format code!

File: Resources/views/Template/shipping.form.html.twig [hunk 21]
```diff
@@ -1,9 +1,9 @@
 <% var disabled = isEditable ? '' : 'disabled' %>
 
-<link rel="stylesheet" type="text/css" href="/bundles/sulusalesshipping/css/main.min.css"/>
+{#<link rel="stylesheet" type="text/css" href="/bundles/sulusalesshipping/css/main.min.css"/>#}
```
Comment (michaelzangerle) [inline Resources/views/Template/shipping.form.html.twig]: @turbo-ele not needed?

File: Resources/views/Template/shipping.form.html.twig [hunk 22]
```diff
@@ -26,163 +27,99 @@
...
-        {#</div>#}
     </form>
 </div>
 
 {% block data %}
     <script type="text/javascript">
-        (function() {
-            App.emit('sulu.salesshipping.set-shipping-status', {{ shippingStatus |json_encode()|raw }});
-        })();
+        {#(function() {#}
```
Comment (michaelzangerle) [inline Resources/views/Template/shipping.form.html.twig]: @turbo-ele should be implemented! the the util/shipingStatus.js could be removed etc.

File: Shipping/ShippingManager.php [hunk 23]
```diff
@@ -376,6 +422,35 @@ private function initializeFieldDescriptors($locale)
                     )
             )
         );
+
+        $this->fieldDescriptors['orderNumber'] = $this->orderFieldDescriptors['orderNumber'];
+
+
+    }
+
+    private function initializeOrderFieldDescriptors($locale) {
```
Comment (michaelzangerle) [inline Shipping/ShippingManager.php]: @turbo-ele unused param

File: Shipping/ShippingManager.php [hunk 24]
```diff
@@ -10,7 +10,14 @@
```
Comment (michaelzangerle) [inline Shipping/ShippingManager.php:10]: @turbo-ele format code

**Comment by michaelzangerle** [comment]
@turbo-ele strange bug with min and max values for shipping item

**Pull request merged**